js/common/modules/@arangodb/foxx/check-args.js

```diff
@@ -207,64 +242,90 @@ exports.validateStatus = function (value) {
   return {value: status, error: null};
 };
 
-exports.validateSchema = function (value) {
-  if (value) {
-    if (value.isJoi || typeof value.validate === 'function') {
-      return {value, error: null};
+exports.validateSchema = function (schema = true) {
+  if (schema === null) {
+    return {
+      value: {schema: false, validate: createSchemaValidator(false)},
+      error: null
+    };
+  }
+  if (isJoiSchema(schema)) {
+    if (schema.isJoi) {
+      return {
+        value: {schema, validate: createSchemaValidator(schema)},
+        error: null
+      };
     }
     try {
-      return {value: joi.object(value).required(), error: null};
+      schema = joi.object(schema).required();
     } catch (e) {
-      return {value, error: Object.assign(
-          new Error('"value" must be a schema'),
-          {cause: e}
+      return {value: schema, error: Object.assign(
+        new Error('"value" must be a schema'),
+        {cause: e}
       )};
     }
+    return {
+      value: {schema, validate: createSchemaValidator(schema)},
+      error: null
+    };
   }
-  return {value, error: new Error('"value" must be a schema.')};
+  const valid = ajv.validateSchema(schema);
+  if (!valid) {
+    return {value: schema, error: new Error(
+      ajv.errorsText(ajv.errors, {dataVar: `"value"`})
+    )};
+  }
+  return {
+    value: {schema, validate: createSchemaValidator(schema)},
+    error: null
+  };
 };
 
 exports.validateModel = function (value) {
-  const warnings = [];
   let model = value;
+  const warnings = [];
+
   let multiple = false;
-  if (model === null) {
-    return {value: {model, multiple}, error: null};
-  }
   if (Array.isArray(model)) {
     if (model.length !== 1) {
       return {value, error: new Error(il`
         "value" must be a model or schema
         or an array containing exactly one model or schema.
-        If you are trying to use multiple schemas
-        try using joi.alternatives instead.
       `), warnings};
     }
     model = model[0];
     multiple = true;
   }
 
-  const index = multiple ? '[0]' : '';
-  if (!model || typeof model !== 'object') {
+  let index = multiple ? '[0]' : '';
+  if (model && typeof model !== "object" && typeof model !== "boolean") {
     return {value, error: new Error(
-        `"value"${index} is not an object`
-    )};
+      `"value"${index} must be an object or boolean`
+    ), warnings};
   }
 
-  const result = exports.validateSchema(model);
-  if (!result.error) {
-    model = {schema: result.value};
+  let schemaIndex = index;
+  if (model && !model.isJoi && Object.keys(model).length && (
+    Object.values(model).some((prop) => typeof prop === "function")
+    || model.schema !== undefined
+  )) {
```

Comment:
This is a bit yucky but we need a way to distinguish between a JS object representing a JSON schema and a JS object that's a model (containing a JOI or JSON schema in a "schema" property). The empty object is a JSON schema and any object with the `isJoi` flag is a JOI schema.